File: android/app/src/main/java/com/rnnotifcations/notifications/Emitter.kt
package com.rnnotifcations.notifications

sealed class RemoteMessagingEvent {
    class NewToken(val token: String) : RemoteMessagingEvent()
    class NewMessage(val message: RemoteMessage) : RemoteMessagingEvent()
}

interface RemoteMessagingSubscriber {
    fun onEvent(event: RemoteMessagingEvent)
}

class RemoteMessagingEventEmitter {
    companion object {
        val shared = RemoteMessagingEventEmitter()
    }

    private val subscribers = mutableListOf<RemoteMessagingSubscriber>()

    fun subscribe(subscriber: RemoteMessagingSubscriber) {
        subscribers.add(subscriber)
    }

    fun unsubscribe(subscriber: RemoteMessagingSubscriber) {
        subscribers.remove(subscriber)
    }

    fun notify(event: RemoteMessagingEvent) {
        subscribers.forEach {
            it.onEvent(event)
        }
    }
}

File: ios/notifications/emitter.swift
import Foundation

enum RemoteMessagingEvent {
  case newToken(String)
  case newMessage(RemoteMessage)
  case notificationOpened(RemoteMessage)
}

protocol RemoteMessagingObserver {
  var id: String { get }
  func onEvent(event: RemoteMessagingEvent)
}

class RemoteMessagingEventEmitter {
  static let shared = RemoteMessagingEventEmitter()
  
  private var observers: [RemoteMessagingObserver] = []
  
  func subscribe(_ emitter: RemoteMessagingObserver) {
    observers.append(emitter)
  }
  
  func unsubscribe(_ emitter: RemoteMessagingObserver) {
    observers = observers.filter { $0.id != emitter.id }
  }
  
  func notify(event: RemoteMessagingEvent) {
    observers.forEach { $0.onEvent(event: event) }
  }
}
